server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,427 @@
+     * maxBytesInFlight
+     */
+    public BulkProcessor2 add(IndexRequest request) throws EsRejectedExecutionException {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds an {@link DeleteRequest} to the list of actions to execute.
+     * @throws EsRejectedExecutionException if adding the approximate size in bytes of the request to totalBytesInFlight would exceed
+     * maxBytesInFlight
+     */
+    public BulkProcessor2 add(DeleteRequest request) throws EsRejectedExecutionException {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds either a delete or an index request.
+     * @throws EsRejectedExecutionException if the total bytes already in flight exceeds maxBytesInFlight. In this case, the request will
+     * not be retried and it is on the client to decide whether to wait and try later.
+     */
+    public BulkProcessor2 add(DocWriteRequest<?> request) throws EsRejectedExecutionException {
+        internalAdd(request);
+        return this;
+    }
+
+    /*
+     * Exposed for unit testing
+     */
+    long getTotalBytesInFlight() {
+        return totalBytesInFlight.get();
+    }
+
+    protected void ensureOpen() {
+        if (closed) {
+            throw new IllegalStateException("bulk process already closed");
+        }
+    }
+
+    private void internalAdd(DocWriteRequest<?> request) throws EsRejectedExecutionException {
+        // bulkRequest and instance swapping is not threadsafe, so execute the mutations under a mutex.
+        // once the bulk request is ready to be shipped swap the instance reference unlock and send the local reference to the handler.
+        Tuple<BulkRequest, Long> bulkRequestToExecute;
+        synchronized (mutex) {
+            ensureOpen();
+            if (totalBytesInFlight.get() >= maxBytesInFlight.getBytes()) {
+                throw new EsRejectedExecutionException(
+                    "Cannot index request of size "
+                        + bulkRequestUnderConstruction.estimatedSizeInBytes()
+                        + " because "
+                        + totalBytesInFlight.get()
+                        + " bytes are already in flight and the max is "
+                        + maxBytesInFlight
+                );
+            }
+            long bytesBeforeNewRequest = bulkRequestUnderConstruction.estimatedSizeInBytes();
+            bulkRequestUnderConstruction.add(request);
+            totalBytesInFlight.addAndGet(bulkRequestUnderConstruction.estimatedSizeInBytes() - bytesBeforeNewRequest);
+            bulkRequestToExecute = newBulkRequestIfNeeded();
+        }
+        // execute sending the local reference outside the lock to allow handler to control the concurrency via it's configuration.
+        if (bulkRequestToExecute != null) {
+            execute(bulkRequestToExecute.v1(), bulkRequestToExecute.v2());
+        }
+        scheduleFlushTask();
```

Comment:
Probably worth a comment about why this line is here (it does makes sense, but I could see it being something that trips somebody up in the future).